forcetree.c

```diff
@@ -27,13 +27,17 @@
  *  reconstruct the tree every timestep.
  */
 
+/*The node index is an integer with unusual properties:
+ * no = 0..All.MaxPart (firstnode in internal functions) corresponds to a particle.
+ * no = All.MaxPart..All.MaxPart + MaxNodes (firstnode..lastnode) corresponds to actual tree nodes,
```

Comment:
So we have internal tree nodes, node-like leaf tree nodes, and real leaf nodes that are particles / pseudo partilces.
To distinguish them, we shall probably call the particles / pseudo particles ParticleNodes.

Currently, the address of the node is used to encode the type of the node. I think we can introduce a struct to clean this up. It is 4 bytes extra but likely worth it (also allows easier removal of particles while preserving the tree structure).
```
typedef union {
   unsigned int IsParticle :1;
   int index :31; /* 2 Billion particle + Pseudo particles at most. */
} NodeRef;

const NodeRef REMOVED_NODE = {.IsParticle  = 1, index = -1}; /* for example */

```

Reply:
If we are going to do this, why not make them real pointers?

typedef union {
   unsigned int IsParticle :1;
   NODE * index :31; /* 2 Billion particle + Pseudo particles at most. */
} NodeRef;

const NodeRef REMOVED_NODE = {.IsParticle  = 1, index = -1}; /* for example */

and then do:
if(IsParticle) (struct particle_data) NodeRef
to access it

Although this is almost the classic case for C++ type inheritance really.

Reply:
Pointer requires special care for serialization. Indices are machine independent. 

I think this is different from classical c++ type inheritance -- the type-specific data here is floating; while in C++ the subclass data is embedded. We manage the type-specific data together. More or less like an placement new in C++ ...